app/controllers/categories_controller.rb

```diff
@@ -0,0 +1,78 @@
+class CategoriesController < ApplicationController
+
+  before_action :require_login
+
+  def index
+    @categories = Category.all
+    @user = User.find(session[:user_id])
+  end
+
+  def new
+    user = User.find(session[:user_id])
+    if user.admin.eql?(1)
+      @category = Category.new
+    else
+      flash[:error] = "You are not authorized to view this page"
+      render file: 'public/404.html.erb'
+    end
+  end
+
+  def show
+    @category = Category.find(params[:id])
+    render :show
+  end
+
+  def edit
+    user = User.find(session[:user_id])
+    if user.admin.eql?(1)
```

Comment:
since you're doing this chunk in multiple actions, i suggest creating another filter to do this for you so you don't have to rewrite the code!